src/elfcore.cc

```diff
@@ -296,17 +296,17 @@ class SysCalls {
   do {                 \
   } while ((fn) < 0 && ERRNO == EINTR)
 
-/* Since we know the proc path for self is constant we don't want to
- * spend CPU to manipulate the path for self.
+/* This function will append the file name to the proc path then
+ * open the file, and return the fd.
  */
-#define OPEN_PROC_FILE(_fd, _proc_path, _proc_file_ind, _file_name, _ext_proc)      \
-  if(_ext_proc) {                                                 \
-    _proc_path[_proc_file_ind] = '\0';                            \
-    strcat(_proc_path, _file_name);                               \
-    NO_INTR(_fd = sys_open(_proc_path, O_RDONLY, 0));             \
-  } else {                                                        \
-    NO_INTR(_fd = sys_open("/proc/self/"_file_name, O_RDONLY, 0)); \
-  }
+static int OpenProcFile(char *proc_path, int proc_file_ind, const char *file_name) {
+  int fd = -1;
+
+  proc_path[proc_file_ind] = '\0';
+  strcat(proc_path, file_name);
```

Comment:
you can't use strcat() for the self case, right? you need a util open-coded version of it.

Reply:
Yes, based on the command we cannot use libc, but I still see some places in original code calling strcat, I can call the local util strcat here.